### PR TITLE
[Feature] Support BigGAN-Deep Model

### PR DESCRIPTION
**Pull request by plyfager**
1. I test inference precision with the author's and hugging face's implementations.
2. There are some unit tests and docstring that need to be completed and refined.

### CODE REVIEW

**Comment by nbei** [comment]
Please fix the linting error and add unit test for this PR.

File: mmgen/models/architectures/biggan/generator_discriminator_deep.py [hunk 1]
```diff
@@ -0,0 +1,635 @@
...
+            to dict(type='BigGANGenResBlock').
+        arch_cfg (dict, optional): Config for the architecture of this
+            generator. Defaults to None.
+        out_norm_cfg (dict, optional): Config for the norm of output layer.
+            Defaults to dict(type='BN').
+        pretrained (str | dict, optional): Path for the pretrained model or
+            dict containing information for pretained models whose necessary
+            key is 'ckpt_path'. Besides, you can also provide 'prefix' to load
+            the generator part from the whole state dict. Defaults to None.
+        reshape_style (str, optional): The reshape style for first linear
```
Comment (nbei) [inline mmgen/models/architectures/biggan/generator_discriminator_deep.py:79]: Why do we need this flag? If this flag is for adopting pre-trained models, you can just re-orgnize the converted weights.

File: mmgen/models/architectures/biggan/generator_discriminator_deep.py [hunk 2]
```diff
@@ -0,0 +1,635 @@
...
+        self.reshape_style = reshape_style
+
+        # Validity Check
+        # If 'num_classes' equals to zero, we shall set 'with_shared_embedding'
+        # to False.
+        if num_classes == 0:
+            assert not self.with_shared_embedding
+            assert not self.concat_noise
+        else:
+            if not self.with_shared_embedding:
```
Comment (nbei) [inline mmgen/models/architectures/biggan/generator_discriminator_deep.py:132]: It seems you can directly use `elif`

File: mmgen/models/architectures/biggan/generator_discriminator_deep.py [hunk 3]
```diff
@@ -0,0 +1,635 @@
...
+                # If not `with_shared_embedding`, we will use `nn.Embedding` to
+                # replace the original `Linear` layer in conditional BN.
+                # Meanwhile, we do not adopt split noises.
+                assert not self.concat_noise
```
Comment (nbei) [inline mmgen/models/architectures/biggan/generator_discriminator_deep.py:136]: What's the relation between `concat_noise` and `split noises`?

File: mmgen/models/architectures/biggan/generator_discriminator_deep.py [hunk 4]
```diff
@@ -0,0 +1,635 @@
...
+                input_is_label=(num_classes > 0)
+                and (not with_shared_embedding),
+                with_spectral_norm=with_spectral_norm,
+                auto_sync_bn=auto_sync_bn))
+
+        self.conv_blocks = nn.ModuleList()
+        for index, out_ch in enumerate(self.arch['out_channels']):
+            for depth in range(self.block_depth):
+                # change args to adapt to current block
+                self.blocks_cfg.update(
```
Comment (nbei) [inline mmgen/models/architectures/biggan/generator_discriminator_deep.py:185]: It is very dangerous to directly update the `self.blocks_cfg`. You may use:

```python
block_cfg_  = deepcopy(self.blocks_cfg)
block_cfg_.updtea(xxx)
```

File: mmgen/models/architectures/biggan/generator_discriminator_deep.py [hunk 5]
```diff
@@ -0,0 +1,635 @@
...
+        if self.concat_noise:
+            if class_vector is not None:
+                z = torch.cat([noise_batch, class_vector], dim=1)
+                y = z
+        elif self.num_classes > 0:
+            z = noise_batch
+            y = class_vector
+        else:
+            z = noise_batch
+            y = noise_batch
```
Comment (nbei) [inline mmgen/models/architectures/biggan/generator_discriminator_deep.py:339]: `y=None`? It's confusing to assign `y` with `noise_batch`.

File: mmgen/models/architectures/biggan/generator_discriminator_deep.py [hunk 6]
```diff
@@ -0,0 +1,635 @@
...
+
+        # First linear layer
+        x = self.noise2feat(z)
+        # Reshape
+        if self.reshape_style == 'tf':
```
Comment (nbei) [inline mmgen/models/architectures/biggan/generator_discriminator_deep.py:344]: Can this part be removed?

File: mmgen/models/architectures/biggan/generator_discriminator_deep.py [hunk 7]
```diff
@@ -0,0 +1,635 @@
...
+            x = x.permute(0, 3, 1, 2).contiguous()
+        else:
+            x = x.view(x.size(0), -1, self.input_scale, self.input_scale)
+        # Loop over blocks
+        for idx, conv_block in enumerate(self.conv_blocks):
+            # Second inner loop in case block has multiple layers
+            if isinstance(conv_block, SelfAttentionBlock):
+                x = conv_block(x)
+            else:
+                x = conv_block(x, y)
```
Comment (nbei) [inline mmgen/models/architectures/biggan/generator_discriminator_deep.py:355]: In some cases, y will be `noise_batch`. Will it work correctly in these cases?

**Review by nbei — COMMENTED** [review]
Comments for the generator.

File: mmgen/models/architectures/biggan/generator_discriminator_deep.py [hunk 8]
```diff
@@ -0,0 +1,635 @@
...
+            number. Defaults to 96.
+        block_depth (int, optional): The repeat times of Residual Blocks in
+            each level of architecture. Defaults to 2.
+        sn_eps (float, optional): Epsilon value for spectral normalization.
+            Defaults to 1e-6.
+        init_type (str, optional): The name of an initialization method:
+            ortho | N02 | xavier. Defaults to 'ortho'.
+        act_cfg (dict, optional): Config for the activation layer.
+            Defaults to dict(type='ReLU').
+        conv_cfg (dict, optional): Config for the convolution module used in
```
Comment (nbei) [inline mmgen/models/architectures/biggan/generator_discriminator_deep.py:433]: This argument is a little confusing and may be removed if we have never changed it.

File: mmgen/models/architectures/biggan/generator_discriminator_deep.py [hunk 9]
```diff
@@ -0,0 +1,635 @@
...
+            this discriminator. Defaults to dict(type='Conv2d').
+        with_spectral_norm (bool, optional): Whether to use spectral
+            normalization. Defaults to True.
+        blocks_cfg (dict, optional): Config for the convolution block.
+            Defaults to dict(type='BigGANDiscResBlock').
+        arch_cfg (dict, optional): Config for the architecture of this
```
Comment (nbei) [inline mmgen/models/architectures/biggan/generator_discriminator_deep.py:439]: After reading the description here, I cannot know how to write the config here. You may specify more about this argument.

File: mmgen/models/architectures/biggan/generator_discriminator_deep.py [hunk 10]
```diff
@@ -0,0 +1,635 @@
...
+            conv_cfg=conv_cfg,
+            with_spectral_norm=with_spectral_norm,
+            spectral_norm_cfg=dict(eps=sn_eps),
+            act_cfg=None)
+
+        self.conv_blocks = nn.ModuleList()
+        for index, out_ch in enumerate(self.arch['out_channels']):
+            for depth in range(self.block_depth):
+                # change args to adapt to current block
+                self.blocks_cfg.update(
```
Comment (nbei) [inline mmgen/models/architectures/biggan/generator_discriminator_deep.py:493]: Directly updating `sefl.blocks_cfg` is dangerous.

File: mmgen/models/architectures/biggan/generator_discriminator_deep.py [hunk 11]
```diff
@@ -0,0 +1,635 @@
...
+
+        Args:
+            x (torch.Tensor): Fake or real image tensor.
+            label (torch.Tensor | None): Label Tensor. Defaults to None.
+
+        Returns:
+            torch.Tensor: Prediction for the reality of the input image with
+                given label.
+        """
+        x0 = self.input_conv(x)
```
Comment (nbei) [inline mmgen/models/architectures/biggan/generator_discriminator_deep.py:582]: Why do you adopt `x0` here? It seems that you can keep using `x` instead of `x0`.

File: mmgen/models/architectures/biggan/modules.py [hunk 12]
```diff
@@ -364,7 +364,7 @@ def __init__(self,
                 kernel_size=1,
                 stride=1,
                 padding=0,
-                conv_cfg=conv_cfg,
+                conv_cfg=shortcut_cfg,
```
Comment (nbei) [inline mmgen/models/architectures/biggan/modules.py:367]: Is this necessary?

File: mmgen/models/architectures/biggan/modules.py [hunk 13]
```diff
@@ -432,3 +432,316 @@ def forward(self, x):
...
+        upsample_cfg (dict, optional): Config for the upsampling operation.
+            Defaults to dict(type='nearest', scale_factor=2).
+        sn_eps (float, optional): Epsilon value for spectral normalization.
+            Defaults to 1e-6.
+        bn_eps (float, optional): Epsilon value for batch normalization.
+            Defaults to 1e-5.
+        with_spectral_norm (bool, optional): Whether to use spectral
+            normalization in this block. Defaults to True.
+        input_is_label (bool, optional): Whether the input of BNs' linear layer
+            is raw label. Defaults to False.
```
Comment (nbei) [inline mmgen/models/architectures/biggan/modules.py:459]: xxx is raw labels **instead of one-hot labels**. This makes the definition of raw labels clear.

File: mmgen/models/architectures/biggan/modules.py [hunk 14]
```diff
@@ -432,3 +432,316 @@ def forward(self, x):
...
+        is_head_block (bool, optional): Whether this block is the first block
+            of BigGAN. Defaults to False.
+    """
+
+    def __init__(self,
+                 in_channels,
+                 out_channels,
+                 channel_ratio=4,
+                 conv_cfg=dict(type='Conv2d'),
+                 shortcut_cfg=dict(type='Conv2d'),
```
Comment (nbei) [inline mmgen/models/architectures/biggan/modules.py:637]: Is it necessary?

File: mmgen/models/architectures/biggan/generator_discriminator_deep.py (hunk 3 above)
Comment (plyfager) [reply mmgen/models/architectures/biggan/generator_discriminator_deep.py:136]: Will, In BigGAN-Deep, if we set `concat_noise` to True, the noise will be concatenated with class vector and then passed to the linear layer and ResBlocks. Otherwise, the noise will be passed to the linear layer alone, and the one-hot label will be passed to the ResBlocks. In BigGAN, if we set 'split_noise' to True, the noise will be split into slots and sent to the linear layer and ResBlock(concatenated with class vector) separately. Otherwise, the noise will be sent to the linear layer only. In the official PyTorch [implementation](https://github.com/ajbrock/BigGAN-PyTorch), both arguments are named with `hier`.

File: mmgen/models/architectures/biggan/generator_discriminator_deep.py (hunk 4 above)
Comment (plyfager) [reply mmgen/models/architectures/biggan/generator_discriminator_deep.py:185]: The same mistakes appear in biggan's `generator_discriminator`, I will fix both.

File: mmgen/models/architectures/biggan/generator_discriminator_deep.py (hunk 5 above)
Comment (plyfager) [reply mmgen/models/architectures/biggan/generator_discriminator_deep.py:339]: If `num_classes` is set to 0, `noise_batch` will be sent to the linear layer and ResBlocks. I just don't want to use the `use_cbn` argument, now it seems that I have to change my `BigGANConditionBN` to adapt to the case when `y=None`.

File: mmgen/models/architectures/biggan/generator_discriminator_deep.py (hunk 7 above)
Comment (plyfager) [reply mmgen/models/architectures/biggan/generator_discriminator_deep.py:355]: In fact, my unit tests cover this case. However, I will change the logic when `num_classes==0`. In other words, `y` would be `None` when `num_classes` is set to 0.

File: mmgen/models/architectures/biggan/generator_discriminator_deep.py (hunk 8 above)
Comment (plyfager) [reply mmgen/models/architectures/biggan/generator_discriminator_deep.py:433]: Which argument? If you mean `conv_cfg`, I will remove it soon.

File: mmgen/models/architectures/biggan/generator_discriminator_deep.py (hunk 9 above)
Comment (plyfager) [reply mmgen/models/architectures/biggan/generator_discriminator_deep.py:439]: I will tell users how to build their own models at the beginning of this docstring.

File: mmgen/models/architectures/biggan/generator_discriminator_deep.py (hunk 11 above)
Comment (plyfager) [reply mmgen/models/architectures/biggan/generator_discriminator_deep.py:582]: There are two paths for the input, using `x0` and `x` to represent them clearly.

File: mmgen/models/architectures/biggan/generator_discriminator_deep.py (hunk 9 above)
Comment (nbei) [reply mmgen/models/architectures/biggan/generator_discriminator_deep.py:439]: Ok

File: mmgen/models/architectures/biggan/generator_discriminator_deep.py (hunk 8 above)
Comment (nbei) [reply mmgen/models/architectures/biggan/generator_discriminator_deep.py:433]: Yep, it's `conv_cfg`.

File: mmgen/models/architectures/biggan/generator_discriminator_deep.py (hunk 7 above)
Comment (nbei) [reply mmgen/models/architectures/biggan/generator_discriminator_deep.py:355]: Ok, thanks.

File: mmgen/models/architectures/biggan/generator_discriminator_deep.py (hunk 11 above)
Comment (nbei) [reply mmgen/models/architectures/biggan/generator_discriminator_deep.py:611]: In fact, for the input `x`, there is only one path here. Isn't it?

File: mmgen/models/architectures/biggan/modules.py [hunk 15]
```diff
@@ -431,3 +425,298 @@ def forward(self, x):
...
+            to the hidden channels' number. Defaults to 4.
+        act_cfg (dict, optional): Config for the activation layer. Defaults to
+            dict(type='ReLU', inplace=False).
+        sn_eps (float, optional): Epsilon value for spectral normalization.
+            Defaults to 1e-6.
+        with_downsample (bool, optional): Whether to use downsampling in this
+            block. Defaults to True.
+        with_spectral_norm (bool, optional): Whether to use spectral
+            normalization. Defaults to True.
+        is_head_block (bool, optional): Whether this block is the first block
```
Comment (nbei) [inline mmgen/models/architectures/biggan/modules.py:610]: Seems to miss this argument in the signature.
